src/main/java/minijava/ir/optimize/ConstantControlFlowOptimizer.java

```diff
@@ -38,15 +53,229 @@ public void visit(Cond node) {
+        // We redirect by replacing predecessor nodes by a BAD.
+        Node bad = graph.newBad(Mode.getANY());
+        phi.setPred(i, bad);
+        currentBlock.setPred(i, bad);
+
+        //Dump.dumpGraph(graph, "after-reach-def");
+      }
+    }
+  }
+
+  /**
+   * Returns a definition for {@param preferredDef} that is actually visible at {@param usage},
+   * defaulting to {@param defaultDefinition} where it is not.
+   */
+  private Node joinReachableDefinitions(
+      Node preferredDef, Node defaultDefinition, Block usage, PSet<Block> loopBreakers) {
+    //System.out.println(
+    //    "    Trying to find "
+    //        + preferredDef
+    //        + ", defaulting to "
+    //        + defaultDefinition
+    //        + ", from usage in "
+    //        + usage);
+    assert preferredDef.getMode() == defaultDefinition.getMode();
+
+    if (Dominance.dominates((Block) preferredDef.getBlock(), usage)) {
+      // hooray, we found our preferred definition!
+      //System.out.println("    Found " + preferredDef + " for " + usage);
+      return preferredDef;
+    }
+
+    if (defaultDefinition.getBlock().equals(usage)) {
+      // We followed edges until we found the default definition, which is consequently deemed visible.
+      //System.out.println("    Reached default definition " + defaultDefinition);
+      return defaultDefinition;
+    }
+
+    if (loopBreakers.contains(usage)) {
+      // We follow some kind of endless loop. Just default to the defaultDefinition
+      // and pray that it's visible (e.g. dominates the usage).
+      assert Dominance.dominates((Block) defaultDefinition.getBlock(), usage);
+      //System.out.println(
+      //    "    Loop detected. Defaulting to default definition " + defaultDefinition);
+      return defaultDefinition;
+    }
+
+    // Otherwise, we have to ask our predecessors and join results with Phi
+    int n = usage.getPredCount();
+    assert n > 0;
+    Node[] preds = new Node[n];
+    boolean allSame = true;
+    for (int i = 0; i < n; ++i) {
+      preds[i] =
+          joinReachableDefinitions(
+              preferredDef,
+              defaultDefinition,
+              (Block) usage.getPred(i).getBlock(),
+              loopBreakers.plus(usage));
+      if (i > 0) {
+        allSame &= preds[0].equals(preds[i]);
+      }
+    }
+
+    if (allSame) {
+      // We don't need to create a new phi node for this one
+      //System.out.println("    Forwarding allSame definition " + preds[0]);
+      return preds[0];
+    }
+
+    // TODO: Possibly pass a hashmap for memoization of nodes.
+    // This can blow up quadratically otherwise, I think.
+    Node newPhi = graph.newPhi(usage, preds, preferredDef.getMode());
+    //System.out.println("    Combined definitions into " + newPhi + " of block " + usage);
+    return newPhi;
+  }
+
+  private static List<Phi> getPhiNodes(Block block) {
+    return seq(BackEdges.getOuts(block)).map(be -> be.node).ofType(Phi.class).toList();
+  }
+
```

Comment:
I tried hard to name stuff in a way that makes sense, but you can only get so far with naming things, I'm afraid... At some point you need to drop comments as to *why* the code what it does, so that's what the comments are for. 

It looks really complicated, because it actually is, much more so than I hoped it would be. Most complexity comes from the fact that we have to deal with the changed dominance frontiers due to the control flow redirects, which basically invalidates some usages.